File: Game-1-modular/Combat/combat_data_loader.py
"""
Combat data loader — dynamic attack generation.

Instead of loading weapon attacks, enemy attacks, and projectiles from JSON,
this module generates them dynamically from weapon properties, enemy definitions,
and tags. This follows the project philosophy: system mechanics in code,
content values from existing JSON (items, enemies).

Attack definitions are created on-the-fly based on:
- Weapon type, range, attack speed, and tags
- Enemy category, tier, tags, and special abilities
- Projectile visuals derived from tags (no projectile JSON)
"""

import math
import random
from typing import Dict, List, Optional, Any

from Combat.attack_state_machine import AttackDefinition
from Combat.projectile_system import ProjectileDefinition
from Combat.hitbox_system import HitboxDefinition


# ============================================================================
# TAG-DRIVEN ATTACK GENERATION
# ============================================================================

# Weapon type -> base attack shape/timing profile
_WEAPON_PROFILES = {
    'sword_1h': {'shape': 'arc', 'arc': 90, 'windup': 300, 'active': 200, 'recovery': 240},
    'sword_2h': {'shape': 'arc', 'arc': 120, 'windup': 500, 'active': 300, 'recovery': 400},
    'dagger':   {'shape': 'arc', 'arc': 60, 'windup': 160, 'active': 120, 'recovery': 160},
    'axe':      {'shape': 'arc', 'arc': 100, 'windup': 400, 'active': 240, 'recovery': 360},
    'mace':     {'shape': 'arc', 'arc': 80, 'windup': 440, 'active': 260, 'recovery': 400},
    'hammer_2h': {'shape': 'arc', 'arc': 130, 'windup': 600, 'active': 360, 'recovery': 500},
    'spear':    {'shape': 'line', 'length': 2.5, 'windup': 360, 'active': 200, 'recovery': 300},
    'staff':    {'shape': 'arc', 'arc': 70, 'windup': 400, 'active': 240, 'recovery': 320},
    'bow':      {'shape': 'projectile', 'windup': 600, 'active': 100, 'recovery': 400},
    'unarmed':  {'shape': 'arc', 'arc': 70, 'windup': 200, 'active': 160, 'recovery': 200},
}

# Element tag -> telegraph/visual color
ELEMENT_COLORS = {
    "physical": [220, 220, 240],
    "fire": [255, 120, 30],
    "ice": [100, 200, 255],
    "frost": [100, 200, 255],
    "lightning": [255, 255, 80],
    "poison": [100, 255, 80],
    "arcane": [180, 80, 255],
    "shadow": [130, 80, 180],
    "holy": [255, 255, 180],
}

# Enemy category -> base attack profile
_ENEMY_ATTACK_PROFILES = {
    'beast':     {'shape': 'arc', 'arc': 80, 'windup': 600, 'active': 240, 'recovery': 400},
    'ooze':      {'shape': 'circle', 'radius': 1.0, 'windup': 800, 'active': 400, 'recovery': 600},
    'insect':    {'shape': 'arc', 'arc': 60, 'windup': 400, 'active': 160, 'recovery': 300},
    'construct': {'shape': 'arc', 'arc': 100, 'windup': 800, 'active': 400, 'recovery': 600},
    'undead':    {'shape': 'arc', 'arc': 90, 'windup': 700, 'active': 300, 'recovery': 500},
    'elemental': {'shape': 'circle', 'radius': 1.2, 'windup': 700, 'active': 360, 'recovery': 500},
    'aberration': {'shape': 'arc', 'arc': 120, 'windup': 600, 'active': 320, 'recovery': 400},
    'dragon':    {'shape': 'cone', 'arc': 140, 'windup': 1000, 'active': 500, 'recovery': 700},
    'humanoid':  {'shape': 'arc', 'arc': 80, 'windup': 500, 'active': 200, 'recovery': 360},
}


def _color_from_tags(tags: List[str], fallback=None) -> List[int]:
    """Get element color from tags."""
    if fallback is None:
        fallback = [220, 220, 240]
    for t in tags:
        if t in ELEMENT_COLORS:
            return list(ELEMENT_COLORS[t])
    return list(fallback)


def generate_weapon_attack(weapon_type: str, weapon_range: float,
                           attack_speed: float = 1.0,
                           weapon_tags: Optional[List[str]] = None) -> AttackDefinition:
    """Generate an attack definition dynamically from weapon properties.

    The hitbox radius matches the actual weapon range so visuals align with mechanics.
    """
    tags = weapon_tags or []
    profile = _WEAPON_PROFILES.get(weapon_type, _WEAPON_PROFILES['unarmed'])

    # Scale timing by attack speed (faster = shorter phases)
    speed_factor = 1.0 / max(0.3, attack_speed)

    # Determine shape and params based on profile
    shape = profile['shape']
    hitbox_params = {'offset_forward': 0.8}

    if shape == 'arc':
        hitbox_params['shape'] = 'arc'
        hitbox_params['radius'] = weapon_range  # Match actual weapon range
        hitbox_params['arc_degrees'] = profile['arc']
    elif shape == 'line':
        hitbox_params['shape'] = 'line'
        hitbox_params['length'] = weapon_range  # Match actual weapon range
    elif shape == 'projectile':
        hitbox_params['shape'] = 'arc'
        hitbox_params['radius'] = 0.5  # Small for projectile spawn point
        hitbox_params['arc_degrees'] = 30

    telegraph_color = _color_from_tags(tags)

    return AttackDefinition(
        attack_id=f"dynamic_{weapon_type}",
        windup_ms=profile['windup'] * speed_factor,
        active_ms=profile['active'] * speed_factor,
        recovery_ms=profile['recovery'] * speed_factor,
        cooldown_ms=100 * speed_factor,
        hitbox_shape=hitbox_params.get('shape', 'arc'),
        hitbox_params=hitbox_params,
        damage_multiplier=1.0,
        movement_multiplier=0.7,
        can_be_interrupted=True,
        animation_id=f"swing_{weapon_type}",
        projectile_id=f"dynamic_{weapon_type}_proj" if shape == 'projectile' else None,
        status_tags=[t for t in tags if t in ('burn', 'bleed', 'poison', 'freeze', 'stun', 'slow')],
        screen_shake=False,
        telegraph_color=telegraph_color,
        tags=tags,
    )


def generate_enemy_attack(enemy_def, attack_index: int = 0) -> AttackDefinition:
    """Generate an attack definition from an enemy definition.

    Uses the enemy's category for base shape, tier for timing/radius scaling,
    and tags for visual coloring.
    """
    category = getattr(enemy_def, 'category', 'beast')
    tier = getattr(enemy_def, 'tier', 1)
    enemy_tags = getattr(enemy_def, 'tags', [])
    visual_size = getattr(enemy_def, 'visual_size', 1.0)

    profile = _ENEMY_ATTACK_PROFILES.get(category, _ENEMY_ATTACK_PROFILES['beast'])

    # Tier scales attack radius and timing
    tier_radius_mult = {1: 1.0, 2: 1.3, 3: 1.6, 4: 2.0}.get(tier, 1.0)
    tier_speed_mult = {1: 1.0, 2: 0.95, 3: 0.9, 4: 0.85}.get(tier, 1.0)

    shape = profile['shape']
    hitbox_params = {'offset_forward': visual_size * 0.6}

    base_radius = visual_size * 0.8 * tier_radius_mult

    if shape == 'arc' or shape == 'cone':
        hitbox_params['shape'] = 'arc'
        hitbox_params['radius'] = base_radius
        hitbox_params['arc_degrees'] = profile.get('arc', 90)
    elif shape == 'circle':
        hitbox_params['shape'] = 'circle'
        hitbox_params['radius'] = base_radius
    elif shape == 'line':
        hitbox_params['shape'] = 'line'
        hitbox_params['length'] = base_radius * 1.5

    telegraph_color = _color_from_tags(enemy_tags, [255, 100, 100])

    return AttackDefinition(
        attack_id=f"enemy_{getattr(enemy_def, 'enemy_id', 'unknown')}_{attack_index}",
        windup_ms=profile['windup'] * tier_speed_mult,
        active_ms=profile['active'] * tier_speed_mult,
        recovery_ms=profile['recovery'] * tier_speed_mult,
        cooldown_ms=200 * tier_speed_mult,
        hitbox_shape=hitbox_params.get('shape', 'arc'),
        hitbox_params=hitbox_params,
        damage_multiplier=1.0,
        movement_multiplier=0.5,
        can_be_interrupted=True,
        animation_id=f"enemy_{category}_attack",
        status_tags=[],
        screen_shake=tier >= 3,
        telegraph_color=telegraph_color,
        tags=enemy_tags,
    )


def generate_projectile_from_tags(weapon_type: str,
                                  tags: Optional[List[str]] = None,
                                  base_speed: float = 12.0,
                                  base_range: float = 10.0) -> ProjectileDefinition:
    """Create a projectile definition dynamically from tags.

    Visual shape, color, trail, and behavior are all derived from tags.
    No JSON needed.
    """
    tags = tags or ['physical']
    color = _color_from_tags(tags)

    # Determine visual shape from tags
    is_arrow = any(t in tags for t in ('arrow', 'bow', 'crossbow'))
    is_beam = any(t in tags for t in ('beam', 'lightning'))
    is_shard = any(t in tags for t in ('ice', 'frost', 'crystal'))

    if is_arrow:
        vis_shape = 'elongated'
        length_px, width_px = 12, 3
    elif is_beam:
        vis_shape = 'beam'
        length_px, width_px = 20, 4
    elif is_shard:
        vis_shape = 'elongated'
        length_px, width_px = 8, 5
    else:
        vis_shape = 'orb'
        length_px, width_px = 8, 8

    # Trail type from element tags
    trail_type = None
    for t in tags:
        if t in ('fire', 'ice', 'frost', 'lightning', 'poison', 'arcane', 'shadow', 'holy'):
            trail_type = f"{t}_trail"
            break

    # Glow for magic projectiles
    has_glow = any(t in tags for t in ('fire', 'ice', 'frost', 'lightning', 'arcane', 'holy', 'shadow'))

    visual = {
        'shape': vis_shape,
        'color': color,
        'glow': has_glow,
        'glow_color': [min(255, c + 60) for c in color],
        'length_px': length_px,
        'width_px': width_px,
    }

    # Homing for certain types
    homing = 0.0
    if any(t in tags for t in ('homing', 'seeking')):
        homing = 0.5

    return ProjectileDefinition(
        projectile_id=f"dynamic_{weapon_type}_proj",
        speed=base_speed,
        max_range=base_range,
        hitbox_radius=0.3,
        sprite_id=vis_shape,
        trail_type=trail_type,
        homing=homing,
        gravity=0.0,
        piercing=any(t in tags for t in ('pierce', 'piercing')),
        visual=visual,
        tags=tags,
    )


# ============================================================================
# COMBAT DATA LOADER
# ============================================================================

class CombatDataLoader:
    """Provides combat data via dynamic generation from weapon/enemy properties.

    No JSON files required. All attack definitions are generated from
    weapon type, range, tags, and enemy definitions.
    """

    def __init__(self):
        self._weapon_cache: Dict[str, AttackDefinition] = {}
        self._enemy_cache: Dict[str, List[AttackDefinition]] = {}
        self._projectile_cache: Dict[str, ProjectileDefinition] = {}
        self._loaded = True  # Always "loaded" — data is generated dynamically

    def load_all(self) -> bool:
        """No-op — data is generated dynamically. Returns True."""
        print("✓ Combat data loader ready (dynamic generation)")
        return True

    @property
    def is_loaded(self) -> bool:
        return True

    # --- Weapon attacks ---

    def get_weapon_attack(self, weapon_type: str,
                          combo_index: int = 0,
                          weapon_range: float = 1.5,
                          attack_speed: float = 1.0,
                          weapon_tags: Optional[List[str]] = None) -> AttackDefinition:
        """Get or generate attack definition for a weapon type."""
        cache_key = f"{weapon_type}_{weapon_range}_{attack_speed}"
        if cache_key not in self._weapon_cache:
            self._weapon_cache[cache_key] = generate_weapon_attack(
                weapon_type, weapon_range, attack_speed, weapon_tags)
        return self._weapon_cache[cache_key]

    def get_weapon_combo_length(self, weapon_type: str) -> int:
        """No combo chains — single attacks only."""
        return 1

    def get_attack_by_id(self, attack_id: str) -> Optional[AttackDefinition]:
        """Look up cached attack by ID."""
        for atk in self._weapon_cache.values():
            if atk.attack_id == attack_id:
                return atk
        for enemy_attacks in self._enemy_cache.values():
            for atk in enemy_attacks:
                if atk.attack_id == attack_id:
                    return atk
        return None

    # --- Enemy attacks ---

    def get_enemy_hurtbox_radius(self, enemy_id: str) -> float:
        """Get hurtbox radius — delegates to EnemyDefinition.hurtbox_radius."""
        return 0.5  # Default; actual value comes from enemy.definition.hurtbox_radius

    def get_enemy_attacks(self, enemy_id: str,
                          enemy_def=None) -> List[AttackDefinition]:
        """Get attack definitions for an enemy, generating from its definition."""
        if enemy_id in self._enemy_cache:
            return self._enemy_cache[enemy_id]

        if enemy_def is None:
            # Return a default attack
            default = generate_enemy_attack(type('Def', (), {
                'category': 'beast', 'tier': 1, 'tags': [], 'visual_size': 1.0,
                'enemy_id': enemy_id
            })())
            return [default]

        attacks = [generate_enemy_attack(enemy_def, 0)]
        self._enemy_cache[enemy_id] = attacks
        return attacks

    def get_enemy_phases(self, enemy_id: str) -> Dict[str, Dict[str, float]]:
        """No phase system from JSON — returns empty."""
        return {}

    def select_enemy_attack(self, enemy_id: str,
                            dist_to_player: float,
                            health_ratio: float = 1.0,
                            enemy_def=None) -> Optional[AttackDefinition]:
        """Select an enemy attack based on range."""
        attacks = self.get_enemy_attacks(enemy_id, enemy_def)
        if not attacks:
            return None

        available = []
        for atk in attacks:
            max_reach = atk.hitbox_radius + atk.hitbox_offset_forward
            if atk.projectile_id:
                max_reach = 999
            if dist_to_player <= max_reach * 1.3:
                available.append(atk)

        if not available:
            return None
        return random.choice(available)

    # --- Projectiles ---

    def get_projectile(self, projectile_id: str,
                       weapon_type: str = 'bow',
                       tags: Optional[List[str]] = None) -> ProjectileDefinition:
        """Get or generate a projectile definition from tags."""
        if projectile_id in self._projectile_cache:
            return self._projectile_cache[projectile_id]

        proj = generate_projectile_from_tags(weapon_type, tags)
        self._projectile_cache[projectile_id] = proj
        return proj

    # --- Hitbox definition helper ---

    @staticmethod
    def hitbox_def_from_attack(attack_def: AttackDefinition) -> HitboxDefinition:
        """Create a HitboxDefinition from an AttackDefinition's hitbox_params."""
        params = attack_def.hitbox_params
        return HitboxDefinition(
            shape=attack_def.hitbox_shape,
            radius=params.get("radius", 1.5),
            arc_degrees=params.get("arc_degrees", 90.0),
            width=params.get("width", 1.0),
            height=params.get("height", 0.5),
            length=params.get("length", 2.0),
            offset_forward=params.get("offset_forward", 0.8),
            offset_lateral=params.get("offset_lateral", 0.0),
            piercing=params.get("piercing", False),
        )

    def clear_cache(self):
        """Clear all cached definitions."""
        self._weapon_cache.clear()
        self._enemy_cache.clear()
        self._projectile_cache.clear()


# Module-level singleton
_loader = None


def get_combat_data_loader() -> CombatDataLoader:
    """Get global combat data loader instance."""
    global _loader
    if _loader is None:
        _loader = CombatDataLoader()
    return _loader

File: Game-1-Godot/src/Game1.Core/Combat/CombatDataLoader.cs
namespace Game1.Core.Combat;

/// <summary>
/// Port of Combat/combat_data_loader.py — dynamic attack generation from
/// weapon properties, enemy definitions, and tags. No JSON.
/// Python uses the global `random` module for weighted attack selection;
/// this port takes an explicit PythonRandom so seeded runs replay exactly.
/// </summary>
public static class CombatGen
{
    public static readonly Dictionary<string, (string Shape, double Arc, double Radius, double Windup, double Active, double Recovery)>
        WeaponProfiles = new()
        {
            ["sword_1h"] = ("arc", 55, 0, 300, 200, 240),
            ["sword_2h"] = ("arc", 75, 0, 500, 300, 400),
            ["dagger"] = ("arc", 25, 0, 160, 120, 160),
            ["axe"] = ("arc", 65, 0, 400, 240, 360),
            ["mace"] = ("arc", 50, 0, 440, 260, 400),
            ["hammer_2h"] = ("arc", 75, 0, 600, 360, 500),
            ["spear"] = ("arc", 15, 0, 360, 200, 300),
            ["staff"] = ("arc", 30, 0, 400, 240, 320),
            ["bow"] = ("projectile", 0, 0, 600, 100, 400),
            ["unarmed"] = ("arc", 50, 0, 200, 160, 200),
        };

    public static readonly Dictionary<string, int[]> ElementColors = new()
    {
        ["physical"] = new[] { 230, 230, 255 },
        ["fire"] = new[] { 255, 100, 20 },
        ["ice"] = new[] { 80, 200, 255 },
        ["frost"] = new[] { 80, 200, 255 },
        ["lightning"] = new[] { 255, 255, 60 },
        ["poison"] = new[] { 80, 255, 60 },
        ["arcane"] = new[] { 200, 60, 255 },
        ["shadow"] = new[] { 160, 60, 220 },
        ["holy"] = new[] { 255, 255, 160 },
    };

    public static readonly Dictionary<string, (string Shape, double Arc, double Radius, double Windup, double Active, double Recovery)>
        EnemyAttackProfiles = new()
        {
            ["beast"] = ("arc", 80, 0, 600, 240, 400),
            ["ooze"] = ("circle", 0, 1.0, 800, 400, 600),
            ["insect"] = ("arc", 60, 0, 400, 160, 300),
            ["construct"] = ("arc", 100, 0, 800, 400, 600),
            ["undead"] = ("arc", 90, 0, 700, 300, 500),
            ["elemental"] = ("circle", 0, 1.2, 700, 360, 500),
            ["aberration"] = ("arc", 120, 0, 600, 320, 400),
            ["dragon"] = ("cone", 140, 0, 1000, 500, 700),
            ["humanoid"] = ("arc", 80, 0, 500, 200, 360),
        };

    private static readonly string[] WeaponStatusTags =
        { "burn", "bleed", "poison", "freeze", "stun", "slow" };

    public static List<int> ColorFromTags(IReadOnlyList<string> tags, int[]? fallback = null)
    {
        fallback ??= new[] { 220, 220, 240 };
        foreach (var t in tags)
            if (ElementColors.TryGetValue(t, out var c))
                return new List<int>(c);
        return new List<int>(fallback);
    }

    public static AttackDefinition GenerateWeaponAttack(
        string weaponType, double weaponRange,
        double attackSpeed = 1.0, List<string>? weaponTags = null)
    {
        var tags = weaponTags ?? new List<string>();
        var profile = WeaponProfiles.TryGetValue(weaponType, out var p)
            ? p : WeaponProfiles["unarmed"];

        var speedFactor = 1.0 / Math.Max(0.3, attackSpeed);

        var shape = profile.Shape;
        var hp = new HitboxParams { OffsetForward = 0.8 };

        if (shape == "arc")
        {
            hp.Shape = "arc";
            hp.Radius = weaponRange;
            hp.ArcDegrees = profile.Arc;
        }
        else if (shape == "line")
        {
            hp.Shape = "line";
            hp.Length = weaponRange;
        }
        else if (shape == "projectile")
        {
            hp.Shape = "arc";
            hp.Radius = 0.5;
            hp.ArcDegrees = 30;
        }

        var telegraphColor = ColorFromTags(tags);

        return new AttackDefinition
        {
            AttackId = $"dynamic_{weaponType}",
            WindupMs = profile.Windup * speedFactor,
            ActiveMs = profile.Active * speedFactor,
            RecoveryMs = profile.Recovery * speedFactor,
            CooldownMs = 100 * speedFactor,
            HitboxShape = hp.Shape ?? "arc",
            HitboxParamsData = hp,
            DamageMultiplier = 1.0,
            MovementMultiplier = 0.7,
            CanBeInterrupted = true,
            AnimationId = $"swing_{weaponType}",
            ProjectileId = shape == "projectile" ? $"dynamic_{weaponType}_proj" : null,
            StatusTags = tags.Where(t => WeaponStatusTags.Contains(t)).ToList(),
            ScreenShake = false,
            TelegraphColor = telegraphColor,
            Tags = tags,
        };
    }

    /// <summary>CPython random.choices(pop, weights, k=1)[0]: cumulative
    /// weights + bisect_right(cum, random()*total, 0, n-1).</summary>
    public static T WeightedChoice<T>(IReadOnlyList<T> population,
                                      IReadOnlyList<double> weights,
                                      PythonRandom rng)
    {
        var n = population.Count;
        var cum = new double[n];
        double acc = 0;
        for (var i = 0; i < n; i++)
        {
            acc += weights[i];
            cum[i] = acc;
        }
        var total = cum[n - 1];
        var u = rng.NextDouble() * total;
        // bisect_right over cum[0..hi) with hi = n-1
        int lo = 0, hi = n - 1;
        while (lo < hi)
        {
            var mid = (lo + hi) / 2;
            if (u < cum[mid]) hi = mid;
            else lo = mid + 1;
        }
        return population[lo];
    }

    public static AttackDefinition GenerateEnemyAttack(
        EnemyDefinition enemyDef, int attackIndex, PythonRandom rng)
    {
        var category = enemyDef.Category;
        var tier = (int)enemyDef.Tier;
        var enemyTags = enemyDef.Tags;
        var visualSize = enemyDef.VisualSize;

        var tierSpeedMult = tier switch { 1 => 1.0, 2 => 0.95, 3 => 0.9, 4 => 0.85, _ => 1.0 };

        var perEnemyAttacks = enemyDef.Attacks;
        if (perEnemyAttacks.Count > 0)
        {
            var weights = perEnemyAttacks.Select(a => (double)a.Weight).ToList();
            var selected = WeightedChoice(perEnemyAttacks, weights, rng);

            var shape = selected.Shape;
            var hp = new HitboxParams { OffsetForward = visualSize * 0.6 };

            if (shape is "arc" or "cone")
            {
                hp.Shape = "arc";
                hp.Radius = selected.Range;
                hp.ArcDegrees = selected.Arc;
            }
            else if (shape == "circle")
            {
                hp.Shape = "circle";
                hp.Radius = selected.Range;
            }
            else if (shape == "line")
            {
                hp.Shape = "line";
                hp.Length = selected.Range;
            }
            else if (shape == "rect")
            {
                hp.Shape = "rect";
                hp.Width = selected.Range * 0.5;
                hp.Height = selected.Range;
            }

            // Python `selected.tags or enemy_tags` — empty list is falsy
            var atkTags = selected.Tags.Count > 0 ? selected.Tags : enemyTags;
            var telegraphColor = ColorFromTags(atkTags, new[] { 255, 100, 100 });

            return new AttackDefinition
            {
                AttackId = $"enemy_{enemyDef.EnemyId}_{selected.AttackId}",
                WindupMs = selected.Windup * tierSpeedMult,
                ActiveMs = selected.Active * tierSpeedMult,
                RecoveryMs = selected.Recovery * tierSpeedMult,
                CooldownMs = 200 * tierSpeedMult,
                HitboxShape = hp.Shape ?? "arc",
                HitboxParamsData = hp,
                DamageMultiplier = selected.DamageMultiplier,
                MovementMultiplier = 0.5,
                CanBeInterrupted = true,
                AnimationId = $"enemy_{category}_{selected.AttackId}",
                StatusTags = new List<string>(selected.StatusTags),
                ScreenShake = selected.ScreenShake,
                TelegraphColor = telegraphColor,
                Tags = new List<string>(atkTags),
            };
        }

        // Fallback: category-based profile
        var profile = EnemyAttackProfiles.TryGetValue(category, out var pr)
            ? pr : EnemyAttackProfiles["beast"];
        var tierRadiusMult = tier switch { 1 => 1.0, 2 => 1.3, 3 => 1.6, 4 => 2.0, _ => 1.0 };

        var fShape = profile.Shape;
        var fhp = new HitboxParams { OffsetForward = visualSize * 0.6 };
        var baseRadius = visualSize * 0.8 * tierRadiusMult;

        if (fShape is "arc" or "cone")
        {
            fhp.Shape = "arc";
            fhp.Radius = baseRadius;
            fhp.ArcDegrees = profile.Arc == 0 ? 90 : profile.Arc;
        }
        else if (fShape == "circle")
        {
            fhp.Shape = "circle";
            fhp.Radius = baseRadius;
        }
        else if (fShape == "line")
        {
            fhp.Shape = "line";
            fhp.Length = baseRadius * 1.5;
        }

        var fallbackColor = ColorFromTags(enemyTags, new[] { 255, 100, 100 });

        return new AttackDefinition
        {
            AttackId = $"enemy_{enemyDef.EnemyId}_{attackIndex}",
            WindupMs = profile.Windup * tierSpeedMult,
            ActiveMs = profile.Active * tierSpeedMult,
            RecoveryMs = profile.Recovery * tierSpeedMult,
            CooldownMs = 200 * tierSpeedMult,
            HitboxShape = fhp.Shape ?? "arc",
            HitboxParamsData = fhp,
            DamageMultiplier = 1.0,
            MovementMultiplier = 0.5,
            CanBeInterrupted = true,
            AnimationId = $"enemy_{category}_attack",
            StatusTags = new List<string>(),
            ScreenShake = tier >= 3,
            TelegraphColor = fallbackColor,
            Tags = new List<string>(enemyTags),
        };
    }

    public static ProjectileDefinition GenerateProjectileFromTags(
        string weaponType, List<string>? tags = null,
        double baseSpeed = 12.0, double baseRange = 10.0)
    {
        tags ??= new List<string> { "physical" };
        var color = ColorFromTags(tags);

        var isArrow = tags.Any(t => t is "arrow" or "bow" or "crossbow");
        var isBeam = tags.Any(t => t is "beam" or "lightning");
        var isShard = tags.Any(t => t is "ice" or "frost" or "crystal");

        string visShape;
        int lengthPx, widthPx;
        if (isArrow) { visShape = "elongated"; lengthPx = 8; widthPx = 3; }
        else if (isBeam) { visShape = "beam"; lengthPx = 24; widthPx = 8; }
        else if (isShard) { visShape = "elongated"; lengthPx = 8; widthPx = 5; }
        else { visShape = "orb"; lengthPx = 8; widthPx = 8; }

        string? trailType = null;
        foreach (var t in tags)
        {
            if (t is "fire" or "ice" or "frost" or "lightning" or "poison"
                or "arcane" or "shadow" or "holy")
            {
                trailType = $"{t}_trail";
                break;
            }
        }

        var hasGlow = tags.Any(t => t is "fire" or "ice" or "frost" or "lightning"
            or "arcane" or "holy" or "shadow");

        var visual = new Dictionary<string, object?>
        {
            ["shape"] = visShape,
            ["color"] = color,
            ["glow"] = hasGlow,
            ["glow_color"] = color.Select(c => Math.Min(255, c + 60)).ToList(),
            ["length_px"] = lengthPx,
            ["width_px"] = widthPx,
        };

        var homing = tags.Any(t => t is "homing" or "seeking") ? 0.5 : 0.0;

        return new ProjectileDefinition
        {
            ProjectileId = $"dynamic_{weaponType}_proj",
            Speed = baseSpeed,
            MaxRange = baseRange,
            HitboxRadius = 0.3,
            SpriteId = visShape,
            TrailType = trailType,
            Homing = homing,
            Gravity = 0.0,
            Piercing = tags.Any(t => t is "pierce" or "piercing"),
            Visual = visual,
            Tags = tags,
        };
    }
}

/// <summary>combat_data_loader.py CombatDataLoader — cached dynamic generation.
/// The weapon cache key deliberately excludes weapon_tags (Python quirk:
/// first call's tags win for a given type/range/speed).</summary>
public sealed class CombatDataLoader
{
    private readonly Dictionary<(string, double, double), AttackDefinition> _weaponCache = new();
    private readonly Dictionary<string, List<AttackDefinition>> _enemyCache = new();
    private readonly Dictionary<string, ProjectileDefinition> _projectileCache = new();

    public AttackDefinition GetWeaponAttack(string weaponType,
                                            int comboIndex = 0,
                                            double weaponRange = 1.5,
                                            double attackSpeed = 1.0,
                                            List<string>? weaponTags = null)
    {
        var key = (weaponType, weaponRange, attackSpeed);
        if (!_weaponCache.TryGetValue(key, out var atk))
        {
            atk = CombatGen.GenerateWeaponAttack(weaponType, weaponRange, attackSpeed, weaponTags);
            _weaponCache[key] = atk;
        }
        return atk;
    }

    public List<AttackDefinition> GetEnemyAttacks(string enemyId,
                                                  EnemyDefinition? enemyDef,
                                                  PythonRandom rng)
    {
        if (_enemyCache.TryGetValue(enemyId, out var cached))
            return cached;

        if (enemyDef is null)
        {
            // Python's default stub sets visual_size=1.0; beast/T1 computes 1.0 too
            var stub = new EnemyDefinition
            {
                EnemyId = enemyId, Name = enemyId, Tier = 1, Category = "beast",
                Behavior = "aggressive",
                Ai = new AiPattern("wander", true, true, 0.0, 0.0, false, new List<string>()),
            };
            return new List<AttackDefinition> { CombatGen.GenerateEnemyAttack(stub, 0, rng) };
        }

        var attacks = new List<AttackDefinition> { CombatGen.GenerateEnemyAttack(enemyDef, 0, rng) };
        _enemyCache[enemyId] = attacks;
        return attacks;
    }

    /// <summary>combat_data_loader.py select_enemy_attack — range-gated pool,
    /// then random.choice.</summary>
    public AttackDefinition? SelectEnemyAttack(string enemyId,
                                               double distToPlayer,
                                               EnemyDefinition? enemyDef,
                                               PythonRandom rng)
    {
        var attacks = GetEnemyAttacks(enemyId, enemyDef, rng);
        if (attacks.Count == 0) return null;

        var available = new List<AttackDefinition>();
        foreach (var atk in attacks)
        {
            var maxReach = atk.HitboxRadius + atk.HitboxOffsetForward;
            if (atk.ProjectileId is not null)
                maxReach = 999;
            if (distToPlayer <= maxReach * 1.3)
                available.Add(atk);
        }

        if (available.Count == 0) return null;
        return rng.Choice(available);
    }

    public ProjectileDefinition GetProjectile(string projectileId,
                                              string weaponType = "bow",
                                              List<string>? tags = null)
    {
        if (_projectileCache.TryGetValue(projectileId, out var cached))
            return cached;

        var proj = CombatGen.GenerateProjectileFromTags(weaponType, tags);
        _projectileCache[projectileId] = proj;
        return proj;
    }

    /// <summary>hitbox_def_from_attack — HitboxDefinition from an
    /// AttackDefinition's params with Python .get defaults.</summary>
    public static HitboxDefinition HitboxDefFromAttack(AttackDefinition attackDef)
    {
        var p = attackDef.HitboxParamsData;
        return new HitboxDefinition
        {
            Shape = attackDef.HitboxShape,
            Radius = p.Radius ?? 1.5,
            ArcDegrees = p.ArcDegrees ?? 90.0,
            Width = p.Width ?? 1.0,
            Height = p.Height ?? 0.5,
            Length = p.Length ?? 2.0,
            OffsetForward = p.OffsetForward,
            OffsetLateral = p.OffsetLateral ?? 0.0,
            Piercing = p.Piercing ?? false,
        };
    }

    public void ClearCache()
    {
        _weaponCache.Clear();
        _enemyCache.Clear();
        _projectileCache.Clear();
    }
}
